cache/estargz.go

```diff
@@ -4,83 +4,289 @@ import (
-			a := make(map[string]string)
+		}, func(ctx context.Context) (map[string]string, error) {
 			mu.Lock()
-			bInfo := bInfo
-			for k, v := range annotations {
-				a[k] = v
-			}
+			cInfo := cInfo
 			mu.Unlock()
-			info, err := cs.Info(ctx, bInfo.compressedDigest)
+			if cInfo == nil {
+				mu.Lock()
+				writeErr := writeErr
+				mu.Unlock()
+				if writeErr != nil {
+					return nil, errors.Wrapf(writeErr, "cannot finalize due to write error")
+				}
+				return nil, errors.Errorf("cannot finalize (reason unknown)")
+			}
+
+			// Fill necessary labels
+			info, err := cs.Info(ctx, cInfo.compressedDigest)
 			if err != nil {
 				return nil, errors.Wrap(err, "failed to get info from content store")
 			}
+			var fields []string
 			if info.Labels == nil {
 				info.Labels = make(map[string]string)
 			}
-			info.Labels[containerdUncompressed] = bInfo.uncompressedDigest.String()
-			if _, err := cs.Update(ctx, info, "labels."+containerdUncompressed); err != nil {
+			info.Labels[containerdUncompressed] = cInfo.uncompressedDigest.String()
+			fields = append(fields, "labels."+containerdUncompressed)
+
+			info.Labels[labelEStargzMetadataDigest] = cInfo.metadataDigest.String()
+			info.Labels["containerd.io/gc.ref.content.metadata"] = cInfo.metadataDigest.String()
+			fields = append(fields, "labels."+labelEStargzMetadataDigest, "labels.containerd.io/gc.ref.content.metadata")
+
+			if cInfo.originalTOCDigest != "" {
+				info.Labels[labelEStargzOriginalTOCDigest] = cInfo.originalTOCDigest.String()
+				info.Labels["containerd.io/gc.ref.content.original-toc-digest"] = cInfo.originalTOCDigest.String()
+				fields = append(fields, "labels."+labelEStargzOriginalTOCDigest, "labels.containerd.io/gc.ref.content.original-toc-digest")
+			}
+
+			if _, err := cs.Update(ctx, info, fields...); err != nil {
 				return nil, err
 			}
-			a[containerdUncompressed] = bInfo.uncompressedDigest.String()
+
+			// Fill annotations
+			a := make(map[string]string)
+			a[estargz.TOCJSONDigestAnnotation] = cInfo.tocDigest.String()
+			a[estargz.StoreUncompressedSizeAnnotation] = fmt.Sprintf("%d", cInfo.uncompressedSize)
+			a[containerdUncompressed] = cInfo.uncompressedDigest.String()
 			return a, nil
 		}
 }
 
+type compressionInfo struct {
+	blobInfo
+	metadataDigest    digest.Digest
+	originalTOCDigest digest.Digest
+	tocDigest         digest.Digest
+}
+
+func writeEStargz(ctx context.Context, cs content.Store, src io.Reader, dest io.Writer) (*compressionInfo, error) {
+	tmpID := identity.NewID()
+
+	// Prepare tar-split
+	mw, err := cs.Writer(ctx, content.WithRef(fmt.Sprintf("write-tar-metadata-%s", tmpID)))
+	if err != nil {
+		return nil, err
+	}
+	defer mw.Close()
+	if err := mw.Truncate(0); err != nil { // Old written data possibly remains
+		return nil, err
+	}
+	p := storage.NewJSONPacker(mw)
+	tocW, err := cs.Writer(ctx, content.WithRef(fmt.Sprintf("write-toc-tar-of-%s", tmpID)))
```

Comment:
That's not quite what I had in mind. Instead of writing out the headers separately my idea was that the stargz encoder would write and compress the original header instead of decoding it. Then uncompressing the stargz the original tar headers should be restored and you need to remove the discard the additional toc that is written into there by the decoder. This standalone object does not persist when pulled from the registry and things still go out of sync.

Reply:
Instead of https://github.com/containerd/stargz-snapshotter/blob/main/estargz/estargz.go#L748 write the `RawBytes` from tar-split parser.